<a href="https://colab.research.google.com/github/techshot25/MachineLearningPractice/blob/master/BareBoneKNN_classifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### This is my first attempt to write a simple classifier using KNN

---

First I have to define training data using the data sets for iris (found at https://en.wikipedia.org/wiki/Iris_flower_data_set)

The $\vec{x}$ and $y$ represent the input and output and $f(x)$ represents the classifier function. The input is a multivariable array of inputs and $y$ would be our single variable classifier (or output).   This is a simple code that takes the nearest neighbor using training data and uses that to predict the output. A more sophisticated method may be used for accuracy but this is just to practice writing the classifier itself.


This is an example of $f(\vec{x}_i)=y_i$ where each $\vec{x}_i$ is an array of some input variables. 

In [20]:
from scipy.spatial import distance

def euc(a,b):
  return distance.euclidean(a,b)

class bareboneKNN():
  def fit(self, x_train, y_train):
    self.x_train = x_train
    self.y_train = y_train
    
  def predict(self, x_test, y_test):
    predictions = []
    for row in x_test:
      label = self.closest(row)
      predictions.append(label)
    return predictions
  
#single closest neighbor test

  def closest(self,row):
    best_dist = euc(row, self.x_train[0])
    best_index = 0
    for i in range(1, len(self.x_train)):
      dist = euc(row, self.x_train[i])
      if dist < best_dist:
        best_dist = dist
        best_index = i
    return self.y_train[best_index]
      
# using iris datasets for practice
from sklearn import datasets
iris = datasets.load_iris()

x = iris.data
y = iris.target

# split the data into training data and testing data
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.5)

my_classifier = bareboneKNN()
my_classifier.fit(x_train,y_train)
predictions = my_classifier.predict(x_test,y_test)
print(predictions)

from sklearn.metrics import accuracy_score

print(accuracy_score(y_test,predictions))

[1, 2, 1, 0, 2, 0, 0, 0, 2, 1, 0, 2, 1, 0, 1, 1, 1, 1, 0, 1, 1, 2, 0, 2, 0, 2, 0, 1, 2, 2, 0, 0, 1, 2, 2, 2, 1, 0, 2, 0, 0, 1, 1, 0, 1, 2, 2, 2, 1, 2, 0, 0, 0, 1, 1, 2, 2, 1, 1, 2, 2, 0, 1, 1, 0, 2, 0, 1, 0, 2, 2, 2, 2, 0, 2]
0.92


92% accuracy is pretty good for a very simple classifier